python/ray/data/tests/test_actor_pool_map_operator.py

```diff
@@ -556,6 +594,113 @@ def __call__(self, x):
+    ray.init()
+
+    @ray.remote(num_cpus=0)
+    class Signal:
+        def __init__(self):
+            self._node_id = ray.get_runtime_context().get_node_id()
+            self._num_alive_actors = 0
+            self._all_nodes_removed = False
+            self._all_nodes_restarted = False
+
+        async def notify_actor_alive(self):
+            self._num_alive_actors += 1
+
+        async def wait_for_actors_alive(self, value):
+            while self._num_alive_actors != value:
+                await asyncio.sleep(0.01)
+
+        async def notify_nodes_removed(self):
+            self._all_nodes_removed = True
+
+        async def notify_nodes_restarted(self):
+            self._all_nodes_restarted = True
+
+        async def wait_for_nodes_removed(self):
+            while not self._all_nodes_removed:
+                await asyncio.sleep(0.01)
+
+        async def wait_for_nodes_restarted(self):
+            while not self._all_nodes_restarted:
+                await asyncio.sleep(0.01)
+
+    # Create the signal actor on the head node.
+    signal_actor = Signal.remote()
+
+    # Spin up nodes
+    num_nodes = 1
+    nodes = []
+    for _ in range(num_nodes):
+        nodes.append(cluster.add_node(num_cpus=10))
+    cluster.wait_for_nodes()
+
+    class MyUDF:
+        def __init__(self, signal_actor):
+            self._node_id = ray.get_runtime_context().get_node_id()
+            self._signal_actor = signal_actor
+            self._signal_sent = False
+
+        def __call__(self, batch):
+            if not self._signal_sent:
+                # Notify the Actor is alive
+                self._signal_actor.notify_actor_alive.remote()
+
+                # Wait for the driver to remove nodes. This makes sure all
+                # actors are running tasks when removing nodes.
+                ray.get(self._signal_actor.wait_for_nodes_removed.remote())
+
+                # Wait for the driver to add nodes.
+                ray.get(self._signal_actor.wait_for_nodes_restarted.remote())
+
+                self._signal_sent = True
+
+            return batch
+
+    res = []
+    num_items = 100
+
+    def run_dataset():
+        nonlocal res
+
+        ds = ray.data.range(num_items, override_num_blocks=num_items)
+        ds = ds.map_batches(
+            MyUDF,
+            fn_constructor_args=[signal_actor],
+            concurrency=num_nodes,
+            batch_size=1,
+        )
+        res = ds.take_all()
+
+    # Kick off Actors
+    thread = threading.Thread(target=run_dataset)
+    thread.start()
+
+    # Wait for all actors to start
+    ray.get(signal_actor.wait_for_actors_alive.remote(num_nodes))
+
+    # Remove all the nodes
+    for node in nodes:
+        cluster.remove_node(node)
+    nodes.clear()
+    ray.get(signal_actor.notify_nodes_removed.remote())
+
+    # Add back all the nodes
+    for _ in range(num_nodes):
+        nodes.append(cluster.add_node(num_cpus=10))
+    cluster.wait_for_nodes()
```

Comment:
Wait, if we're adding all of the nodes back, this test should work even without the changes to `pick_actor`, right?

Reply:
yes, but we didn't have such an test case before. adding this would be useful

Reply:
Ok, we'd have just added that test separately as it confused me into thinking that this verifies the changes made in this PR, which it's not